Feature/daniele/checkout custom parameters

DanieleAlessandra opened the pull request:
This PR introduces support for additional checkout configuration properties in the SDK initializer.

What’s new

The following properties are now recognized during initialization:
```
// currency
'currency'                      => 'string',
'default_currency'              => 'string',
// cart
'always_show_renewals_amount'   => 'bool',
'annual_discount'               => 'bool',
'billing_cycle'                 => ['string', 'int'],
'bundle_discount'               => 'float',
'maximize_discounts'            => 'bool',
'multisite_discount'            => ['bool', 'string'], // string expected to be "auto"
'show_inline_currency_selector' => 'bool',
'show_monthly'                  => 'bool',
// appearance
'form_position'                 => 'string',
'is_bundle_collapsed'           => 'bool',
'layout'                        => 'string',
'refund_policy_position'        => 'string',
'show_refund_badge'             => 'bool',
'show_reviews'                  => 'bool',
'show_upsells'                  => 'bool',
'title'                         => 'string',
```

Behavior
	•	If these properties exist in the initialization config, they are validated and stored.
	•	When opening the checkout (both iframe and redirect modes), the provided values are automatically passed along to the checkout context.

This allows developers to customize the checkout flow appearance and cart behavior more flexibly directly from the SDK initialization.

DanieleFreemius committed:
Added checkout customization parameters

# Conflicts:
#	start.php

DanieleFreemius committed:
Flat checkout properties

fajardoleo commented on includes/class-freemius.php:
```diff
@@ -365,6 +365,15 @@ class Freemius extends Freemius_Abstract {
+        protected $_checkout_config = [];
+
+        /**
+
```
@DanieleAlessandra An extra line.

fajardoleo commented on includes/class-freemius.php:
```diff
@@ -5322,6 +5334,88 @@ private function parse_settings( &$plugin_info ) {
+         *
+         * @return array
+         */
+        protected function validate_checkout_config($config)
```
@DanieleAlessandra Missing spaces around `$config`.

fajardoleo commented on includes/class-freemius.php:
```diff
@@ -5322,6 +5334,88 @@ private function parse_settings( &$plugin_info ) {
+
+            $result = array();
+
+            foreach ($schema as $key => $expected_type)
```
@DanieleAlessandra Please check the spaces.

fajardoleo commented on includes/class-freemius.php:
```diff
@@ -5322,6 +5334,88 @@ private function parse_settings( &$plugin_info ) {
+
+            foreach ($schema as $key => $expected_type)
+            {
+                if (array_key_exists($key, $config))
```
@DanieleAlessandra  Spaces

fajardoleo commented on includes/class-freemius.php:
```diff
@@ -5322,6 +5334,88 @@ private function parse_settings( &$plugin_info ) {
+                        switch ($type)
+                        {
+                            case 'bool':
+                                if (is_bool($value))
```
@DanieleAlessandra Curly braces

fajardoleo commented on includes/class-freemius.php:
```diff
@@ -5322,6 +5334,88 @@ private function parse_settings( &$plugin_info ) {
+         * @return array
+         */
+        public function get_checkout_config()
+        {
```
@DanieleAlessandra Curly braces :)

fajardoleo commented on includes/class-freemius.php:
```diff
@@ -5322,6 +5334,88 @@ private function parse_settings( &$plugin_info ) {
+                // cart
+                'always_show_renewals_amount'   => 'bool',
+                'annual_discount'               => 'bool',
+                'billing_cycle'                 => ['string', 'int'],
```
@DanieleAlessandra Please check this out:
https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#declaring-arrays

Long array syntax.

DanieleFreemius committed:
feat(checkout): allow customizing checkout query params via filter

Introduce a namespaced filter to let developers add/modify query params sent to Freemius Checkout.
	•	Add fs_apply_filter( $fs->get_unique_affix(), 'checkout_query_params', ... )

DanieleFreemius committed:
Add allowlist for checkout query parameters in FS_Checkout_Manager

Introduced the private $_allowed_custom_params property to define all supported
custom query parameters for the checkout. After applying `fs_apply_filter()`,
the function now filters `$filtered_params` using `array_intersect_key()` to
remove any unsupported keys before merging. This prevents external filters from
injecting unexpected parameters into the checkout query.

swashata commented on includes/managers/class-fs-checkout-manager.php:
```diff
@@ -12,7 +12,35 @@
+        /**
+         * Allowlist of query parameters for checkout.
+         */
+        private $_allowed_custom_params = array(
```
@DanieleAlessandra this is kinda misleading because we store `string`, `bool` etc yet we don't do the type verification. Also I don't think we need to.

So let's just do this

```php
array(
  'currency' => true,
  'default_currency' => true,
)
```

and keep your array_intersect_key implementation (which is pretty good)

swashata commented on includes/managers/class-fs-checkout-manager.php:
```diff
@@ -153,7 +181,40 @@ public function get_query_params( Freemius $fs, $plugin_id, $plan_id, $licenses
 			);
 
-			return array_merge( $context_params, $_GET, array(
+            /**
```
Remove this doc block please. See that it is already stale. Instead

1. See the `start.php` and see we have list of `Freemius filters collection`, just add the filter there.
2. Later we will update our docs https://freemius.com/help/documentation/wordpress-sdk/filters-actions-hooks/

swashata commented on includes/managers/class-fs-checkout-manager.php:
```diff
@@ -153,7 +181,40 @@ public function get_query_params( Freemius $fs, $plugin_id, $plan_id, $licenses
+             * @param int|mixed $plan_id       The selected plan ID (if any).
+             * @param int|mixed $licenses      The requested number of licenses (if provided).
+             */
+            $filtered_params = fs_apply_filter(
```
Why not

```
$filtered_params = $fs->apply_filters('checkout_query_params', $context_params);
```

swashata replied:
@DanieleAlessandra let's call the filter

```
checkout/configuration
```

Following we already have this https://freemius.com/help/documentation/wordpress-sdk/filters-actions-hooks/#checkoutpurchasedcompleted

swashata reviewed (changes requested):
@DanieleAlessandra this is a very good approach, please see the small tweaks I suggested.

swashata commented on includes/managers/class-fs-checkout-manager.php:
```diff
@@ -12,7 +12,35 @@
+            'show_refund_badge'             => 'bool',
+            'show_reviews'                  => 'bool',
+            'show_upsells'                  => 'bool',
+            'title'                         => 'string',
```
@DanieleAlessandra let's add the `billing_cycle_selector` in the list as well.

DanieleFreemius committed:
Changed filter name to checkout/parameters, code clean and small fixes

swashata reviewed (approved):
Nicely done 👏
